app/worker/worker.go

```diff
@@ -86,56 +91,137 @@ func InitAndRun(
 		_ = unsubscribe()
 	}()
 
-	var msg interface{}
-	sem := make(chan bool, concurrency)
 	quit := make(chan os.Signal, 1)
 	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// Spawn worker pool
+	var wg = new(sync.WaitGroup)
+	wg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		workerID := i
+		go func() {
+			defer func() {
+				wg.Done()
+				wg.Wait()
+				// Make sure context is cancelled when all workers are down
+				cancel()
```

Comment:
If I understand correctly this code, each worker goroutine will Wait on the workergroup and cancel the context. Does it make sense to cancel the context N times, where N is the concurrency? Shouldn't the cancel simply be called after the wg.Wait at the end of the function?

Reply:
The cancel function is reentrant so it doesn't matter how many times it's called, only the first one will count. Calling cancel in this location is done to make sure the main thread is unblocked when no workers remain. Perhaps we should even move the cancel call before the `wg.Wait()` statement so that the service quits as soon as the first worker thread returns?

Reply:
I removed the `wg.Wait()` call. Now when a worker returns it will cancel the context unblocking the main thread (if it hasn't received a signal already) as well as all other workers. The main thread will wait for all workers to finish their current job before returning.